File: python/fastscore/suite/instance.py

from fastscore.live import ActiveSensor

from ..errors import FastScoreError
from ..v1.rest import ApiException

class InstanceBase(object):
    """
    The parent of all FastScore instance classes.
    """

    class ActiveSensorBag(object):
        def __init__(self, inst):
            self._inst = inst

        def ids(self):
            return [ x['id'] for x in self ]

        def __iter__(self):
            try:
                for x in self._inst.swg.active_sensor_list(self._inst.name):
                    yield ActiveSensor(x['id'], x['tap'])
            except Exception as e:
                raise FastScoreError("Cannot retrieve active sensors", caused_by=e)

        def __getitem__(self, tapid):
            try:
                x = self._inst.swg.active_sensor_describe(self._inst.name, tapid)
            except Exception as e:
                if isinstance(e, ApiException) and e.status == 404:
                    raise FastScoreError("Active sensor #%d not found" % tapid)
                else:
                    raise FastScoreError("Cannot retrieve active sensor", caused_by=e)

        def __delitem__(self, tapid):
            return self._inst.uninstall_sensor(tapid)

    def __init__(self, name, api, swg, swg2=None):
        self.name = name
        self.api = api
        self.swg = swg
        self.swg2 = swg2    # REST API v2
        self._active_sensors = InstanceBase.ActiveSensorBag(self)

    @property
    def active_sensors(self):
        """
        A collection of currently installed sensors indexed by id.

        >>> engine = connect.lookup('engine')
        >>> engine.active_sensors.ids()
        [8]
        >>> x = engine.active_sensors[8]
        >>> x.id
        8
        >>> x.tap
        manifold.input.records.count
        >>> x.uninstall()
        >>> engine.active_sensors.ids()
        []

        """
        return self._active_sensors
   
    @property
    def tapping_points(self):
        """
        A list of tapping points supported by the instance.

        >>> mm.tapping_points
        ['sys.memory',... ]

        """
        return self.swg.active_sensor_available(self.name)

    def check_health(self):
        """
        Retrieves version information from the instance. A successful reply indicates
        that the instance is healthy.

        >>> connect.check_health()
        {
          'id': '366e5030-d773-49cb-8b28-9b1b9d173c79',
          'built_on': 'Thu May 11 12:53:39 UTC 2017',
          'release': '1.5'
        }

        """
        try:
            return self.swg.health_get(self.name)
        except Exception as e:
            raise FastScoreError("Cannot retrieve instance info", caused_by=e)

    def get_swagger(self):
        """
        Retrieves the Swagger specification of the API supported by the
        instance.

        >>> connect.get_swagger()
        {u'info':...}

        """
        try:
            return self.swg.swagger_get(self.name)
        except Exception as e:
            raise FastScoreError("Cannot retrieve Swagger specification", caused_by=e)

    def install_sensor(self, sensor):
        return self.swg.active_sensor_attach(self.name, sensor.desc)

    def uninstall_sensor(self, tapid):
        self.swg.active_sensor_detach(self.name, tapid)

File: R/fastscore/R/instance.R
InstanceBase <- R6::R6Class("InstanceBase",
    inherit = ApiClient, # swagger twin
    public = list(
      name = NA,
      api = NA,
      basePath = NA,

      initialize = function(name = NA, api = NA, basePath){

        stopifnot(grepl("://", basePath), grepl("https:", basePath)) # TODO: add FS error message
        #   FastScoreError$new(message = "basePath must be an URL, e.g. https://dashboard:8000")$error_string()
        #   FastScoreError$new(message = "basePath must use HTTPS scheme, e.g. https://dashboard:8000")$error_string()

        self$name <- name
        self$api <- api
        self$basePath <- paste0(basePath, "/",
                                  parse_url(swagger::ApiClient$new()$basePath)$path)
      },

      # @property (???)
      active_sensors = function(){
        #' Currently installed sensors indexed by id
        #' > engine <- Connect$lookup('engine')
        #' > x <- engine$active_sensors()
        #' > str(x)
        #' > x

        tryCatch(
          self$active_sensor_list(self$name),
          error = function(e) FastScoreError$new(
            message = "Unable to retrieve active sensors.",
            caused_by = e$message
            )$error_string()
          )
      },

      # @property (???)
      tapping_points = function(){
        # List of tapping points supported by the instance
        #
        # > mod_man$tapping_points()

        tryCatch(
          self$tapping_points(self$name),

          error = function(e) FastScoreError$new(
            message = "Unable to list tapping points.",
            caused_by = e$message
            )$error_string()
        )
      },
      # calls  v2.yaml method: tapping_points(...)

      check_health = function(){
        # Retrieves version information from the instance.
        # A successful reply indicates that the instance is healthy.
        #
        # > Connect$check_health()

        tryCatch(
          self$health_get(self$name),

          error = function(e) FastScoreError$new(
            message = "Unable to retrieve instance info.",
            caused_by = e$message
            )$error_string()
        )
      },

      get_swagger = function(){
        # Retrieves the Swagger specification of the API
        # supported by the instance.
        #
        # > Connect$get_swagger()

        tryCatch(
          self$swagger_get(self$name),

          error = function(e) FastScoreError$new(
            message = "Unable to retrieve Swagger specification.",
            caused_by = e$message
          )$error_string()
        )
      },

      install_sensor = function(sensor){
        tryCatch(
          self$active_sensor_attach(self$name, sensor$desc),
          # swagger::ConnectApi$active_sensor_attach(...)
          error = function(e) FastScoreError$new(
            message = "Unable to install sensor.",
            caused_by = e$message
          )$error_string()
        )
      },

      uninstall_sensor = function(tapid){
        tryCatch(
          self$active_sensor_detach(self$name, tapid),
          # swagger::ConnectApi$active_sensor_detach(...)

          error = function(e) FastScoreError$new(
            message = "Unable to uninstall sensor.",
            caused_by = e$message
          )$error_string()
        )
      }
    )
)
